cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -1597,6 +1632,21 @@ class ObjectOutputStream final : public io::OutputStream {
     }
     upload_id_ = outcome.GetResult().GetUploadId();
     upload_state_ = std::make_shared<UploadState>();
```

Comment:
Would this create duplicate `upload_state_` if shifting from single-part req?

Reply:
Yes, we would create the upload state twice. (basically resetting it)
I changed it in [1e9e3a4](https://github.com/apache/arrow/pull/41564/commits/1e9e3a476874b9d464517f5f8357b370e315fbed)